### ISSUE TITLE
[BUG] pthread_kill failed when running vernier through Sidekiq

### ISSUE THREAD

**Issue by DimitriosLisenko** [jhawthorn/vernier#51]
Hello! Really excited to use vernier! I gave it a go yesterday by integrating it into a Sidekiq middleware and unfortunately the Sidekiq process is crashing with the following error in the middle of (some) jobs:
```
[BUG] pthread_kill failed
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin23]

-- C level backtrace information -------------------------------------------
/Users/dima/.rvm/rubies/ruby-3.2.2/lib/libruby.3.2.dylib(rb_vm_bugreport+0x9a0) [0x10332f110]
/Users/dima/.rvm/rubies/ruby-3.2.2/lib/libruby.3.2.dylib(rb_bug_without_die+0x15c) [0x1031699e8]
/Users/dima/.rvm/rubies/ruby-3.2.2/lib/libruby.3.2.dylib(rb_bug+0x1c) [0x103446880]
/Users/dima/.rvm/gems/ruby-3.2.2@appstore-rails-7.1/bundler/gems/vernier-0c3f6dd66236/lib/vernier/vernier.bundle(_ZN19GlobalSignalHandler13record_sampleER10LiveSampleP17_opaque_pthread_t+0x64) [0x129e1ef4c]
/Users/dima/.rvm/gems/ruby-3.2.2@appstore-rails-7.1/bundler/gems/vernier-0c3f6dd66236/lib/vernier/vernier.bundle(_ZN13TimeCollector17sample_thread_runEv) [0x129e1ec60]
/Users/dima/.rvm/gems/ruby-3.2.2@appstore-rails-7.1/bundler/gems/vernier-0c3f6dd66236/lib/vernier/vernier.bundle(_ZN13TimeCollector19sample_thread_entryEPv) [0x129e1e578]
/usr/lib/system/libsystem_pthread.dylib(_pthread_start+0x88) [0x189b92034]
```

I tried generating a backtrace from a core dump, though I'm not sure it's more helpful than the above:
```
* thread #71
    frame #0: 0x0000000189b5a0dc libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000189b91cc0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000189a9da40 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000103169ae8 libruby.3.2.dylib`die at error.c:776:5
    frame #4: 0x0000000103446884 libruby.3.2.dylib`rb_bug(fmt=<unavailable>) at error.c:800:5
    frame #5: 0x0000000129e1ef4c vernier.bundle`GlobalSignalHandler::record_sample(this=0x0000000129e2c368, sample=0x00000002cb080f78, pthread_id=0x00000002b1187000) at vernier.cc:1156:17
  * frame #6: 0x0000000129e1ec60 vernier.bundle`TimeCollector::sample_thread_run(this=0x000000029b6ac8d0) at vernier.cc:1239:58
    frame #7: 0x0000000129e1e578 vernier.bundle`TimeCollector::sample_thread_entry(arg=0x000000029b6ac8d0) at vernier.cc:1275:20
    frame #8: 0x0000000189b92034 libsystem_pthread.dylib`_pthread_start + 136
```

As an aside, with YJIT enabled it crashes even faster, but that's a separate issue probably.

Let me know if I can provide any additional information, will be happy to.

**Comment by jhawthorn** [comment]
Thanks for the report! `pthread_kill` failing likely means the original thread exited, which shouldn't happen (though it is something we could probably more safely recover from).

There's been some significant changes, which haven't yet had a release and it's possible that solves this (though I'm not sure as I didn't expect an existing issue). YJIT being enabled also shouldn't be an issue.

**Comment by DimitriosLisenko** [comment]
Got it, thank you! Looking forward to the next release in that case :)

**Comment by jhawthorn** [comment]
Thanks for testing on 0.4.0. I'm not sure why this is happening, my only guess is that this is relate to a thread going away, but we have tests for that and Vernier should be able to see the thread going away and not attempt to profile it anymore. I could (and likely will) make this fail more safely, we could recover from this error, but first I'd like to figure out why it is occurring.

Can you provide any more details about which jobs cause this failure or what might be interesting or particular about your sidekiq workload?

**Comment by jhawthorn** [comment]
I still haven't diagnosed what causes this, but we added a failsafe in https://github.com/jhawthorn/vernier/commit/971f56e144ce3fc6a44032b1923eb57cba09408a and that is released in 0.5.1

That will almost certainly work around this particular error, but if it just turns into a different one please let me know 🙇‍♂

**Comment by jhawthorn** [comment]
Oh! I think I am seeing evidence of this when `thread.kill` is used but ONLY on Ruby 3.2 and on macos

**Comment by DimitriosLisenko** [comment]
Hey John! Sorry for the delay, been trying (and failing) to find a moment to dig a bit deeper into this :)

That does sound promising! We are indeed using Ruby 3.2, I was personally testing on MacOS, and our application does run a bunch of Sidekiq jobs that might use an internal thread pool or other gems that might spawn/kill threads (which I suppose is relatively rare for Sidekiq workloads).

I can try looking out for the specific place it crashes if you'd like - unless you think you have a solution already for the `thread.kill` issue? As an aside, does it look like both Ruby 3.1 and 3.3 are not affected?

**Comment by DimitriosLisenko** [comment]
@jhawthorn Hello, I'm back! We actually just upgraded our app to be running Ruby 3.3 and the latest version of vernier, and I gave profiling Sidekiq a go again. Unfortunately we're still seeing the same issue, but slightly more insidious because while before it would raise a pthread_kill exception, now it just silently fails - the Sidekiq job just starts randomly hanging at some point, and stays stuck until it's restarted (with no logs).

Are there any debugging tools built into vernier that I could use to try get some additional information?

**Comment by DimitriosLisenko** [comment]
OK so the good news is that I actually managed to replicate this in our application in a really simple way.

Here is the Sidekiq job that I created:
```
class VernierTestJob
  include Sidekiq::Worker
  sidekiq_options queue: :worker_background_priority_low
  sidekiq_options retry: false

  def perform
    sleep_seconds = rand(10)
    ::Rails.logger.info("Sleeping for #{sleep_seconds} seconds")
    sleep(sleep_seconds)
    ::Rails.logger.info("Done sleeping for #{sleep_seconds} seconds")
  end
end
```

and then all I do is run the worker, then open the rails console, and spam `VernierTestJob.perform_async`. After a certain point, the logs stop appearing in the worker, and it completely freezes up until it's restarted.

**Comment by DimitriosLisenko** [comment]
Replicated on a brand new Rails project: https://github.com/DimitriosLisenko/blog

Steps:
```
docker compose up # bring up Redis
bundle exec sidekiq -c 25 # bring up Sidekiq
bundle exec rails c # bring up Rails console and keep spamming VernierTestJob.perform_async until Sidekiq logs freeze
```

Example of Sidekiq being frozen - last log line is the following:
```
2024-12-06T17:17:22.363Z pid=95264 tid=1tco class=VernierTestJob jid=ef556c6100a1dc21f6699cbf INFO: start
```

**Comment by jhawthorn** [comment]
This sounds like a different issue than the original one reported.

That does sound like a bug, and I'd like to fix it, but I don't think this is a good way to run Vernier (or any profiler). Recording separate profile for each concurrent job is going to be very messy (probably hurts accuracy) and I don't think any other profiler would let you do it (stackprof only supports one running profile).

I do want to support it, because it's helpful for testing and other edge cases, but it's probably a better idea to only run *one* profile per-process. Vernier allows running a single profile which captures all running threads, so capturing a longer profile is going to be easier and have better data than many very short profiles.

**Comment by jhawthorn** [comment]
I can't reproduce this on either my Linux machine or my M2 mac. On both I was able to `100.times { VernierTestJob.perform_async }`, and eventually had 100 profiles in my tmp/ directory. However I do see that the process uses very significant resources when multiple profiles are being run at once, which makes sense.

> I have also seen scenarios where the process would freeze for like 5 minutes, then continue running some jobs, and then freezing up again.

This suggests also that this might not be a bug but that it's simply too resource intensive on your machine to run N (up to 25?) profiles at once. It's best to run a single profile at a time.

**Comment by DimitriosLisenko** [comment]
Thanks for testing the replication and getting back to me, appreciate it!

You might be right that it's purely a resource issue - my CPU does spike fairly high when running the profilers in parallel. I'll try running a single profiler instance across multiple Sidekiq jobs instead and see how it goes.

**Comment by DimitriosLisenko** [comment]
OK so I have some more information after doing some digging!

What I'm doing right now is I have the following initializer:
```
$ cat config/initializers/vernier_run.rb
$vernier_index = 0
Signal.trap("TRAP") do
  Thread.new do
    if $vernier_index % 2 == 0
      puts "Starting vernier profile"
      ::Vernier.start_profile(out: "tmp/#{::DateTime.now.to_utc.iso8601}.profile.json")
    else
      puts "Stopping vernier profile"
      ::Vernier.stop_profile
    end
    $vernier_index = $vernier_index + 1
  end
end
```
and I'm starting the profiler just before I run one of our long-running Sidekiq jobs.

Doing this, I noticed it's getting stuck on the same place as before when I ran the profilers in parallel. Digging a bit deeper, it specifically stopped working in a place where we were using our old framework of executing commands on a remote server via a really old gem:
```
pid, stdin, stdout, stderr = ::Open4::popen4(to_run)
ignored, status = ::Process::waitpid2(pid)
```

I'm pretty sure this forks the process, which I'm guessing doesn't work well with Vernier.

After replacing that part with our new framework (which doesn't fork), I was able to run one of the jobs I really wanted profiled all the way through. Yay!

**Comment by jhawthorn** [comment]
Ah! Yep. Forking will currently hang. I'm working on it
